runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStream.java

```diff
@@ -0,0 +1,346 @@
+    this.nextBudgetAdjustment = new AtomicReference<>(GetWorkBudget.noBudget());
+    this.pendingResponseBudget = new AtomicReference<>(GetWorkBudget.noBudget());
+  }
+
+  public static GrpcDirectGetWorkStream create(
+      Function<
+              StreamObserver<StreamingGetWorkResponseChunk>,
+              StreamObserver<StreamingGetWorkRequest>>
+          startGetWorkRpcFn,
+      GetWorkRequest request,
+      BackOff backoff,
+      StreamObserverFactory streamObserverFactory,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      int logEveryNStreamFailures,
+      ThrottleTimer getWorkThrottleTimer,
+      Supplier<GetDataStream> getDataStream,
+      Supplier<CommitWorkStream> commitWorkStream,
+      ProcessWorkItem processWorkItemFn) {
+    GrpcDirectGetWorkStream getWorkStream =
+        new GrpcDirectGetWorkStream(
+            startGetWorkRpcFn,
+            request,
+            backoff,
+            streamObserverFactory,
+            streamRegistry,
+            logEveryNStreamFailures,
+            getWorkThrottleTimer,
+            getDataStream,
+            commitWorkStream,
+            processWorkItemFn);
+    getWorkStream.startStream();
+    return getWorkStream;
+  }
+
+  private void sendRequestExtension() {
+    GetWorkBudget adjustment = nextBudgetAdjustment.get();
+    StreamingGetWorkRequest extension =
+        StreamingGetWorkRequest.newBuilder()
+            .setRequestExtension(
+                Windmill.StreamingGetWorkRequestExtension.newBuilder()
+                    .setMaxItems(adjustment.items())
+                    .setMaxBytes(adjustment.bytes()))
+            .build();
+
+    executor()
+        .execute(
+            () -> {
+              try {
+                send(extension);
+              } catch (IllegalStateException e) {
+                // Stream was closed.
+              }
+            });
+
+    synchronized (this) {
+      // Just sent the request extension, reset the nextBudgetAdjustment. This will be set when
+      // adjustBudget is called.
+      nextBudgetAdjustment.set(GetWorkBudget.noBudget());
+    }
+  }
+
+  @Override
+  protected synchronized void onNewStream() {
+    workItemBuffers.clear();
+    long currentInflightItems = inflightMessages.get();
+    long currentInflightBytes = inflightBytes.get();
+    GetWorkBudget currentNextBudgetAdjustment = nextBudgetAdjustment.get();
+    GetWorkBudget.Builder newNextBudgetAdjustmentBuilder = currentNextBudgetAdjustment.toBuilder();
+    if (currentInflightItems > 0) {
+      newNextBudgetAdjustmentBuilder.setItems(
+          currentNextBudgetAdjustment.items() + currentInflightItems);
+    }
+
+    if (currentInflightBytes > 0) {
+      newNextBudgetAdjustmentBuilder.setBytes(
+          currentNextBudgetAdjustment.bytes() + currentInflightBytes);
+    }
+
+    GetWorkBudget newNextBudgetAdjustment = newNextBudgetAdjustmentBuilder.build();
+
+    inflightMessages.set(newNextBudgetAdjustment.items());
```

Comment:
use atomicref to the budget since they are immutable?

Reply:
done